src/middleware/performanceToggles.js

```diff
@@ -0,0 +1,28 @@
+import * as types from "../actions/types";
+
+/**
+ * Performance toggles (reduxState.controls.performanceToggles) represent flags
+ * for which we enable/disable certain functionality in Auspice. These flags
+ * shouldn't be depended on, i.e. Auspice should work just fine without them
+ * (but may be a little slow).
+ */
+
+
+export const performanceToggles = (_store) => (next) => (action) => {
+  let modifiedAction;
+  switch (action.type) {
+    case types.URL_QUERY_CHANGE_WITH_COMPUTED_STATE: /* fallthrough */
+    case types.CLEAN_START: {
+      modifiedAction = {...action};
+      modifiedAction.controls.performanceToggles = calculate(action)
+    }
+  }
+  return next(modifiedAction || action); // send action to other middleware / reducers
+};
+
+function calculate({tree}) {
+  const toggles = new Map();
+  const totalTipCount = tree?.nodes?.[0]?.fullTipCount;
+  toggles.set("skipTreeAnimation", totalTipCount > 4000);
```

Comment:
In canonical redux this would be straightforward as middleware has access to the old state and because we use fat actions it also can see the new state. But as we modify the `nodes` directly (i.e. they're not immutable, largely for performance reasons) we can't leverage this. But we can see old and new `idxOfInViewRootNode` so we could implement the last bit of code you wrote within the middleware. (Would require quite a bit of testing!)